File: Source/Tests/ExcelDna.AddIn.RegistrationSampleVB/ExampleAddIn.vb
Imports ExcelDna.Integration

Public Class ExampleAddIn
    Implements IExcelAddIn

    Public Sub AutoOpen() Implements IExcelAddIn.AutoOpen
        ExcelIntegration.RegisterUnhandledExceptionHandler(Function(ex) "!!! ERROR: " + ex.ToString())

        'PerformDefaultRegistration()

        ' Could add Async too...
    End Sub

    Public Sub AutoClose() Implements IExcelAddIn.AutoClose

    End Sub

End Class

File: Source/Tests/ExcelDna.AddIn.RegistrationSampleFS/ExampleAddIn.fs
namespace ExcelDna.AddIn.RegistrationSampleFS

open System
open System.Linq.Expressions
open ExcelDna.Integration
open ExcelDna.Registration

type ExampleAddIn () =
    interface IExcelAddIn with
        member this.AutoOpen ()  = 
            // The overload selection and delegate conversions performed by F# are not intuitive.
            let paramConvertConfig = ParameterConversionConfiguration()
                                        .AddParameterConversion( 
                                            Func<Type, ExcelParameterRegistration, LambdaExpression>(FsParameterConversions.FsOptionalParameterConversion),
                                            null)

            ExcelRegistration.GetExcelFunctions ()
            |> fun fns -> ParameterConversionRegistration.ProcessParameterConversions (fns, paramConvertConfig)
            |> FsAsyncRegistration.ProcessFsAsyncRegistrations
            |> AsyncRegistration.ProcessAsyncRegistrations
            |> ExcelRegistration.RegisterFunctions
        
        member this.AutoClose () = ()

